### DISCUSSION TITLE
API v4 - Unable to Retrieve Email - Resource not accessible by integration

### DISCUSSION THREAD

**Discussion by c-brooks** [community/community#24774] [Apps, API and Webhooks]
<p>Hi,</p>
<p>I am trying to retrieve information on pull requests, including the author’s email (and the emails of reviewers and commentors). Unfortunately whenever I include the “Email” field in my query I get the error “Resource not accessible by integration”. Removing the Email field removes this error.</p>
<p>Simplified query:</p>
<pre><code>query Github_PullRequests($name: String!, $owner: String!) {
  repository(owner: $owner, name: $name) {
    name
    pullRequests(first: 1) {
      nodes {
        number
        title
        author {
          ... on User {
            login
            email # &lt;- causes problem
            name
          }
        }
      }
      pageInfo {
        hasNextPage
        endCursor
      }
      totalCount
    }
  }
}
</code></pre>
<p>Response:</p>
<pre><code>{
  "data": {
    "repository": {
      "name": "api",
      "pullRequests": {
        "nodes": [
          {
            "number": 1,
            "title": "Update project's namespace",
            "author": null
          }
        ],
        "pageInfo": {
          "hasNextPage": true,
          "endCursor": "Y3Vyc29yOnYyOpHOCxMLQA=="
        },
        "totalCount": 487
      }
    }
  },
  "errors": [
    {
      "type": "FORBIDDEN",
      "path": [
        "repository",
        "pullRequests",
        "nodes",
        0,
        "author",
        "email"
      ],
      "locations": [
        {
          "line": 12,
          "column": 13
        }
      ],
      "message": "Resource not accessible by integration"
    }
  ]
}
</code></pre>
<p>The Permissions I have given (all read-only):</p>
<ul>
<li>Repository contents</li>
<li>Issues</li>
<li>Repository metadata </li>
<li>Pull requests </li>
<li>Organization members </li>
<li>Repository webhooks</li>
</ul>
<p>User Permissions:</p>
<ul>
<li>Emails </li>
</ul>
<p>I couldn’t find a comprehensive list of the permissions and what access they give, so forgive me if it’s a silly error on my part. I have tried the same query using a personal access token and it works fine.</p>

**Comment by AndreaGriffiths11** [comment]
<p>Hi <a class="mention" href="/c-brooks">@c-brooks</a> ,</p>
<p>Thanks for being here! And thank you for your patience while we researched your question. To be able to help you better could you answer the following:</p>
<p>What kind of request is this?: is it server-to-server (using an installation token) or user-to-server (using an OAuth token)?</p>

**Comment by c-brooks** [comment]
<p>Thanks for the reply - sorry, I should have stated in the original post. It’s an installation token for a Github App</p>

**Comment by AndreaGriffiths11** [comment, marked as answer]
<p>Hi <a class="mention" href="/c-brooks">@c-brooks</a>,</p>
<p>Thanks again for your patience, here is the feedback based on our Support teams research:</p>
<p>The reason that request returns the specific error message is because a user’s email is only accessible via user-to-server request (in other words, an OAuth token and not an installation token).</p>
<p>Here’s some more context: when a GitHub App is installed to a repo, it requests repo-specific permissions to the authorizing user. However, a GitHub App can only get user-specific information if the user in question has chosen to be identified by the app as a part of the User Authorization flow described here:</p>
<p><a href="https://developer.github.com/apps/building-github-apps/identifying-and-authorizing-users-for-github-apps/#identifying-and-authorizing-users-for-github-apps" rel="nofollow noopener">https://developer.github.com/apps/building-github-apps/identifying-and-authorizing-users-for-github-apps/#identifying-and-authorizing-users-for-github-apps</a></p>
<p>At the end of the authorization flow, that OAuth token can be used to make a request to the GraphQL API to obtain the <code>email</code> value for some author. However, only <code>email</code>s from author’s who have been authorized by the app can be obtained.</p>
<p>Thank you for your feedback on the documentation. It’s a great point all <a href="https://developer.github.com/apps/building-github-apps/identifying-and-authorizing-users-for-github-apps/#user-to-server-requests" rel="nofollow noopener">user-to-server requests</a> are documented, but there isn’t a specific section in GitHub Developer that would tell integrators which GraphQL fields are accessible by scope. We’ve taken your feedback and passed it along to the appropriate teams. Thanks again for reaching out :grinning:</p>

**Comment by allenap** [comment]
<p>Hi <a class="mention" href="/AndreaGriffiths11">@AndreaGriffiths11</a>,</p>
<p>I’m seeing inconsistency in how the access rules are applied.</p>
<p>I developed a GitHub App with a temporary organisation. A few colleagues joined the organisation as guinea pigs, one of whom hides his email. The following query has worked consistently, returning <code>null</code> for that one member’s email address:</p>
<pre><code class="lang-auto">query($nodeId: ID!, $cursor: String) {
  node(id: $nodeId) {
    ... on Organization {
      membersWithRole(first: 100, after: $cursor) {
        edges {
          hasTwoFactorEnabled
          node {
            id
            email
            name
            login
            createdAt
            updatedAt
          }
          role
        }
        pageInfo {
          endCursor
          hasNextPage
        }
      }
    }
  }
}
</code></pre>
<p>Now I have installed my app into another organisation, and I am consistently getting “Resource not accessible by integration” until I remove the <code>email</code> field from my query. Do you know what might cause the difference in behaviour?</p>
<p>Thanks, Gavin.</p>
